### PR TITLE
release: add tidb 4.0.10 release notes

### PR DESCRIPTION
**Pull request by TomShawn**
### What is changed, added or deleted? (Required)

Add tidb 4.0.10 release notes

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v5.0 (TiDB 5.0 versions)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/4608
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: releases/release-4.0.10.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+
+## 新功能
+
+
+## 改进提升
+
+
+## Bug 修复
+
```
Comment (eurekaka) [inline releases/release-4.0.10.md:21]:
```suggestion
TiDB
- 修复由于并发导致的 batch client 超时问题 [#22336](https://github.com/pingcap/tidb/pull/22336)
- 修复由于并发地自动捕获 SQL 绑定而导致的重复绑定问题 [#22295](https://github.com/pingcap/tidb/pull/22295)
- 当日志级别为 'debug' 时，让 SQL 绑定的自动捕获正确运行 [#22293](https://github.com/pingcap/tidb/pull/22293)
- 当 region 合并正在发生时，正确地释放锁 [#22267](https://github.com/pingcap/tidb/pull/22267)
- 对 datetime 类型的用户变量返回正确的值 [#22143](https://github.com/pingcap/tidb/pull/22143)
- 修复错误使用 Index Merge 访问方式的问题 [#22124](https://github.com/pingcap/tidb/pull/22124)
- 修复由于执行计划缓存导致 TiFlash 报 `wrong precision` 错误的问题 [#21960](https://github.com/pingcap/tidb/pull/21960)
- 修复由于 schema 变更导致的错误结果 [#21596](https://github.com/pingcap/tidb/pull/21596)
- 避免在 `ALTER TABLE` 中不必要的 column flag 更改 [#21474](https://github.com/pingcap/tidb/pull/21474)
- 让包含子查询块别名的 optimizer hint 生效 [#21380](https://github.com/pingcap/tidb/pull/21380)
- 为 IndexHashJoin 和 IndexMergeJoin 生成正确的 optimizer hint [#21020](https://github.com/pingcap/tidb/pull/21020)
```

File: releases/release-4.0.10.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+TiDB 版本：4.0.10
+
+## 兼容性更改
+
+
+## 新功能
+
+
+## 改进提升
+
```
Comment (eurekaka) [inline releases/release-4.0.10.md:18]:
```suggestion
TiDB

- 添加 `txn-entry-size-limit` 配置项，用于限制事务中单个 key-value 记录的大小 [#21843](https://github.com/pingcap/tidb/pull/21843)
```

File: releases/release-4.0.10.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+发版日期：2021 年 1 月 15 日
+
+TiDB 版本：4.0.10
+
+## 兼容性更改
+
+
+## 新功能
+
+
```
Comment (LittleFall) [inline releases/release-4.0.10.md:16]:
```suggestion
+ TiFlash
    - 添加了配置项 `security.redact_info_log`，可以设置将日志中的用户数据脱敏，例如把具体的值替换为 `?`.
    - 支持了枚举类型的无损修改。
    
```

File: releases/release-4.0.10.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    - 修复由于并发地自动捕获 SQL 绑定而导致的重复绑定问题 [#22295](https://github.com/pingcap/tidb/pull/22295)
+    - 当日志级别为 'debug' 时，让 SQL 绑定的自动捕获正确运行 [#22293](https://github.com/pingcap/tidb/pull/22293)
+    - 当 region 合并正在发生时，正确地释放锁 [#22267](https://github.com/pingcap/tidb/pull/22267)
+    - 对 datetime 类型的用户变量返回正确的值 [#22143](https://github.com/pingcap/tidb/pull/22143)
+    - 修复错误使用 Index Merge 访问方式的问题 [#22124](https://github.com/pingcap/tidb/pull/22124)
+    - 修复由于执行计划缓存导致 TiFlash 报 `wrong precision` 错误的问题 [#21960](https://github.com/pingcap/tidb/pull/21960)
+    - 修复由于 schema 变更导致的错误结果 [#21596](https://github.com/pingcap/tidb/pull/21596)
+    - 避免在 `ALTER TABLE` 中不必要的 column flag 更改 [#21474](https://github.com/pingcap/tidb/pull/21474)
+    - 让包含子查询块别名的 optimizer hint 生效 [#21380](https://github.com/pingcap/tidb/pull/21380)
+    - 为 IndexHashJoin 和 IndexMergeJoin 生成正确的 optimizer hint [#21020](https://github.com/pingcap/tidb/pull/21020)
```
Comment (LittleFall) [inline releases/release-4.0.10.md:37]:
```suggestion
    - 为 IndexHashJoin 和 IndexMergeJoin 生成正确的 optimizer hint [#21020](https://github.com/pingcap/tidb/pull/21020)
    
+ TiFlash
    
    - 修复了 TiFlash 解析老版本 TiDB 表结构失败导致 TiFlash 无法启动的问题。
    - 修复了在 RedHat 系统中 TiFlash 会对 cpu time 进行错误处理导致 TiFlash 无法启动的问题。
    - 修复了将配置项 path_realtime_mode 设置为 true 时 TiFlash 无法启动的问题。
    - 修复了当调用三个参数的 `substr` 函数时，返回结果错误的问题。
```

File: releases/release-4.0.10.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+## 兼容性更改
+
+
+## 新功能
+
++ TiFlash
+
+    - 添加了配置项 `security.redact_info_log`，可以设置将日志中的用户数据脱敏，例如把具体的值替换为 `?`
+    - 支持了枚举类型的无损修改
```
Comment (LittleFall) [inline releases/release-4.0.10.md:19]: remove this line.
```suggestion
```

File: releases/release-4.0.10.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    - 避免在 `ALTER TABLE` 中不必要的 column flag 更改 [#21474](https://github.com/pingcap/tidb/pull/21474)
+    - 让包含子查询块别名的 optimizer hint 生效 [#21380](https://github.com/pingcap/tidb/pull/21380)
+    - 为 IndexHashJoin 和 IndexMergeJoin 生成正确的 optimizer hint [#21020](https://github.com/pingcap/tidb/pull/21020)
+
++ TiFlash
+
+    - 修复了 TiFlash 解析老版本 TiDB 表结构失败导致 TiFlash 无法启动的问题。
+    - 修复了在 RedHat 系统中 TiFlash 会对 cpu time 进行错误处理导致 TiFlash 无法启动的问题。
+    - 修复了将配置项 path_realtime_mode 设置为 true 时 TiFlash 无法启动的问题。
+    - 修复了当调用三个参数的 `substr` 函数时，返回结果错误的问题。
```
Comment (LittleFall) [inline releases/release-4.0.10.md:48]:
```suggestion
    - 修复了当调用三个参数的 `substr` 函数时，返回结果错误的问题。
    - 修复了当 tidb 对 enum 枚举进行无损修改时，tiflash 无法读取修改后的值的问题
    
```

File: releases/release-4.0.10.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    - 修复了当 tidb 对 enum 枚举进行无损修改时，tiflash 无法读取修改后的值的问题
```
Comment (LittleFall) [inline releases/release-4.0.10.md:48]: remove periods.
```suggestion
+ TiFlash

    - 修复了 TiFlash 解析老版本 TiDB 表结构失败导致 TiFlash 无法启动的问题
    - 修复了在 RedHat 系统中 TiFlash 会对 cpu time 进行错误处理导致 TiFlash 无法启动的问题
    - 修复了将配置项 path_realtime_mode 设置为 true 时 TiFlash 无法启动的问题
    - 修复了当调用三个参数的 `substr` 函数时，返回结果错误的问题
    - 修复了当 tidb 对 enum 枚举进行无损修改时，tiflash 无法读取修改后的值的问题
```

File: releases/release-4.0.10.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (overvenus) [inline releases/release-4.0.10.md:49]:
```suggestion
# TiCDC

## Improvements

* 为 maxwell 协议默认开启 old value [#1144](https://github.com/pingcap/ticdc/pull/1144)
* 默认启用 unified sorter [#1230](https://github.com/pingcap/ticdc/pull/1230)

## Bug Fixes

* 修复 maxwell 协议的问题，包括 base64 数据输出和将 tso 转换成 unix timestamp [#1173](https://github.com/pingcap/ticdc/pull/1173)
* 修复过期的元数据可能引发新创建的 changefeed 异常的问题 [#1184](https://github.com/pingcap/ticdc/pull/1184)
* 修复在关闭的 notifier 上创建 receiver [#1199](https://github.com/pingcap/ticdc/pull/1199)
* 修复在 etcd 更新缓慢时导致的内存增长问题 [#1227](https://github.com/pingcap/ticdc/pull/1227)
* 修复 `max-batch-size` 不生效的问题 [#1253](https://github.com/pingcap/ticdc/pull/1253)
* 修复清理过期任务信息的问题 [#1280](https://github.com/pingcap/ticdc/pull/1280)
* 修复 MySQL sink 中由于没有调用 rollback 而导致的回收 db conn 卡住的问题 [#1285](https://github.com/pingcap/ticdc/pull/1285)

# TiDB-binlog

## Bug Fixes

* 修复在启用 amend txn 时，Drainer 可能会使用错误 schema 来生成 SQL 的问题 [#1033](https://github.com/pingcap/tidb-binlog/pull/1033)

# BR

## Bug Fixes

* 修复 4.0.9 无法恢复 4.0.8 保存在 GCS 上的备份 [#688](https://github.com/pingcap/br/pull/688)
* 修复在恢复 GCS 上的备份时可能发送的 panic 问题 [#673](https://github.com/pingcap/br/pull/673)

# Dumpling

## Improvements

* 支持检查为定义的参数，支持输出导出的进度
 [#228](https://github.com/pingcap/dumpling/pull/228)

## Bug Fixes

* 修改默认设置的 tidb_mem_quota_query 的行为以避免 TiDB OOM
 [#233](https://github.com/pingcap/dumpling/pull/233)

# TiDB-Lightning

## Improvements

* 支持重试读 S3 遇到的错误 [#533](https://github.com/pingcap/tidb-lightning/pull/533)

## Bug Fixes

* 修复未正确编码 region key 而导致分裂 region 失败问题 [#531](https://github.com/pingcap/tidb-lightning/pull/531)
* 修复可能丢失 create table 失败的错误 [#530](https://github.com/pingcap/tidb-lightning/pull/530)
* 修复使用 TiDB backend 时遇到的 "column count mismatch" 问题 [#535](https://github.com/pingcap/tidb-lightning/pull/535)

```

File: releases/release-4.0.10.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    - 当日志级别为 `'debug'` 时，让 SQL 语句绑定的自动捕获正确运行 [#22293](https://github.com/pingcap/tidb/pull/22293)
+    - 当 Region 合并正在发生时，正确地释放锁 [#22267](https://github.com/pingcap/tidb/pull/22267)
+    - 对 `Datetime` 类型的用户变量返回正确的值 [#22143](https://github.com/pingcap/tidb/pull/22143)
+    - 修复错误使用 Index Merge 访问方式的问题 [#22124](https://github.com/pingcap/tidb/pull/22124)
+    - 修复由于执行计划缓存导致 TiFlash 报 `wrong precision` 错误的问题 [#21960](https://github.com/pingcap/tidb/pull/21960)
+    - 修复由于 schema 变更导致的错误结果 [#21596](https://github.com/pingcap/tidb/pull/21596)
+    - 避免在 `ALTER TABLE` 中不必要的 column flag 更改 [#21474](https://github.com/pingcap/tidb/pull/21474)
+    - 让包含子查询块别名的 optimizer hint 生效 [#21380](https://github.com/pingcap/tidb/pull/21380)
+    - 为 `IndexHashJoin` 和 `IndexMergeJoin` 生成正确的 optimizer hint [#21020](https://github.com/pingcap/tidb/pull/21020)
+
```
Comment (tabokie) [inline releases/release-4.0.10.md:53]: TiKV

    - 修复了 peer 和 ready 之间的错误映射 [#9409](https://github.com/tikv/tikv/pull/9409)
    - 修复一些日志信息在 `security.redact-info-log` 设置为 `true` 时未脱敏的问题 [#9314](https://github.com/tikv/tikv/pull/9314)

File: releases/release-4.0.10.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+发版日期：2021 年 1 月 15 日
+
+TiDB 版本：4.0.10
+
+## 新功能
+
++ TiFlash
+
+    - 添加了配置项 `security.redact_info_log`，可以设置将日志中的用户数据脱敏，例如把具体的值替换为 `?`
+
```
Comment (rleungx) [inline releases/release-4.0.10.md:16]: + PD

    - 添加了配置项 `enable-redact-log`，可以设置将日志中的用户数据脱敏，例如把具体的值替换为 `?  [#3266](https://github.com/pingcap/pd/pull/3266)

File: releases/release-4.0.10.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+## 改进提升
+
++ TiDB
+
+    - 添加 `txn-entry-size-limit` 配置项，用于限制事务中单个 key-value 记录的大小 [#21843](https://github.com/pingcap/tidb/pull/21843)
+
```
Comment (rleungx) [inline releases/release-4.0.10.md:22]: + PD

    - 优化了 store-state-filter 的监控，可以显示更加具体的原因 [#3100](https://github.com/tikv/pd/pull/3100)
    - 更新 `go.etcd.io/bbolt` 依赖至 v1.3.5 [#3331](https://github.com/tikv/pd/pull/3331)

File: releases/release-4.0.10.md (hunk 9 above)
Comment (rleungx) [inline releases/release-4.0.10.md:53]:
```suggestion

+ PD

    - 修复 ID 分配不是单调递增的问题 [#3308](https://github.com/tikv/pd/pull/3308) [#3323](https://github.com/tikv/pd/pull/3323)
    - 修复 PD client 在某些情况下可能卡住的问题 [#3285](https://github.com/pingcap/pd/pull/3285)
```

File: releases/release-4.0.10.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
++ PD
+
+    - 优化了 store-state-filter 的监控，可以显示更加具体的原因 [#3100](https://github.com/tikv/pd/pull/3100)
+    - 更新 `go.etcd.io/bbolt` 依赖至 v1.3.5 [#3331](https://github.com/tikv/pd/pull/3331)
++ Tools
```
Comment (rleungx) [inline releases/release-4.0.10.md:31]:
```suggestion

+ Tools
```

**Review by zanmato1984 — COMMENTED** [review]
TiFlash part LGTM

**Comment by scsldb** [comment]
LGTM

**Comment by TomShawn** [comment]
LGTM

**Review by ran-huang — APPROVED** [review]

**Pull request merged**